interface/ceed-operator.c

```diff
@@ -248,6 +248,166 @@ int CeedCompositeOperatorAddSub(CeedOperator compositeop,
+    Method based approximate inverse. This function obtains the simultaneous
+    diagonalization for the 1D mass and Laplacian operators,
+      M = V^T V, K = V^T S V.
+    The assembled QFunction is used to modify the eigenvalues from simultaneous
+    diagonalization and obtain an approximate inverse of the form
+      V^T S^hat V.
+
+  @param op             CeedOperator to create element inverses
+  @param[out] fdminv    CeedOperator to apply the action of a FDM based inverse
+                          for each element
+  @param[out] qdata     CeedVector to hold qdata for fdminv
+  @param request        Address of CeedRequest for non-blocking completion, else
+                          CEED_REQUEST_IMMEDIATE
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Advanced
+**/
+int CeedOperatorCreateFDMElementInverse(CeedOperator op, CeedOperator *fdminv,
+    CeedRequest *request) {
+  int ierr;
+  Ceed ceed = op->ceed;
+
+  return CeedError(ceed, 1, "Not implemented yet.");
+
+  // Determine active input basis
+  bool interp = false, grad = false;
+  CeedBasis basis;
+  for (CeedInt i=0; i<op->nfields; i++)
+    if (op->inputfields[i] && op->inputfields[i]->vec == CEED_VECTOR_ACTIVE) {
+      basis = op->inputfields[i]->basis;
+      interp = interp || op->qf->inputfields[i]->emode == CEED_EVAL_INTERP;
+      grad = grad || op->qf->inputfields[i]->emode == CEED_EVAL_GRAD;
+    }
+  CeedInt P1d = basis->P1d, Q1d = basis->Q1d, dim = basis->dim,
+          ncomp = basis->ncomp;
+
+  // Build and diagonalize 1D Mass and Laplacian
+  if (!basis->tensorbasis)
+    return CeedError(ceed, 1, "FDMElementInverse only supported for tensor bases");
+  // -- Mass
+  CeedScalar *mass, *work;
+  ierr = CeedMalloc(Q1d*P1d, &work); CeedChk(ierr);
+  for (CeedInt i=0; i<Q1d; i++)
+    for (CeedInt j=0; j<P1d; j++)
+      work[i+j*Q1d] = basis->interp1d[i*P1d+j]*basis->qweight1d[i];
+  ierr = CeedMatrixMultiply(ceed, work, basis->interp1d, mass, P1d, P1d, Q1d);
+  CeedChk(ierr);
+  // -- Laplacian
+  CeedScalar *laplace;
+  for (CeedInt i=0; i<Q1d; i++)
+    for (CeedInt j=0; j<P1d; j++)
+      work[i+j*Q1d] = basis->grad1d[i*P1d+j]*basis->qweight1d[i];
+  ierr = CeedMatrixMultiply(ceed, work, basis->grad1d, laplace, P1d, P1d, Q1d);
+  CeedChk(ierr);
+  ierr = CeedFree(&work); CeedChk(ierr);
+  // -- Diagonalize
+  CeedScalar *x, *lambda;
+  ierr = CeedMalloc(P1d*P1d, &x); CeedChk(ierr);
+  ierr = CeedMalloc(P1d, &lambda); CeedChk(ierr);
+  ierr = CeedSimultaneousDiagonalization(ceed, laplace, mass, x, lambda, P1d);
+  CeedChk(ierr);
+  ierr = CeedFree(&mass); CeedChk(ierr);
+  ierr = CeedFree(&grad); CeedChk(ierr);
+
+  // Assemble QFunction
+  CeedVector assembled;
+  CeedElemRestriction rstr;
+  ierr =  CeedOperatorAssembleLinearQFunction(op, &assembled, &rstr, request);
+  CeedChk(ierr);
+  ierr = CeedElemRestrictionDestroy(&rstr); CeedChk(ierr);
+
+  // Calculate element averages
```

Comment:
Suppose I have a target grid function `f[i][j][k]` that I would like to approximate via the product `x[i]*y[j]*z[k]` (perhaps with each of those factors positive). We can defer for now, but it would be cool to "solve" this problem. I think doing so would be useful in boundary layers, for example.

Reply:
Right, I would like to do much better than a simple average here. The average gets us to where we can run simple problems on simple meshes, but I plan on developing more sophisticated diagonals to handle more complex problems. My next step will be taking `lambda` and changing `I kron I kron lambda + I kron lambda kron I + lambda kron I kron I` (for 3D Laplacian) to  `I kron I kron lambda_z + I kron lambda_y kron I + lambda_x kron I kron I` to handle variation along reference axes.